### PR TITLE
Added checkbox locked choices feature

### PR DESCRIPTION
**Pull request by NivEz**
### Checkbox locked choices feature:

This feature gives the option lock some choices.
Locked choices cannot be removed (with space or left button) and they are merged with the default choices, so they appear as checked choices.
This is similar to giving choices to the user which must be chosen in order to continue with the process (like user agreement). The only option not to choose the locked option is to break the program.
This is useful if you want to make it clear that some choice/s out of the list must be chosen in order to continue.

BTW this package is awesome and I use it for installation tool that I might publish soon (created this PR because I needed this feature in my project).

### CODE REVIEW

**Review by staticdev — DISMISSED** [review]
@NivEz thanks a lot for your contribution. Please check out tests in `tests` folder, we are committed to delivering a high quality code and having unit, integration and acceptance tests are a way to ensure that. Also, I see that current code is failing pre-commit checks, you can run it locally with `pre-commit install` and then `pre-commit run --all-files`.  You can find a bit more details in our [contributing guide](https://github.com/magmax/python-inquirer/blob/main/CONTRIBUTING.md).

**Comment by NivEz** [comment]
> @NivEz thanks a lot for your contribution. Please check out tests in `tests` folder, we are committed to delivering a high quality code and having unit, integration and acceptance tests are a way to ensure that. Also, I see that current code is failing pre-commit checks, you can run it locally with `pre-commit install` and then `pre-commit run --all-files`. You can find a bit more details in our [contributing guide](https://github.com/magmax/python-inquirer/blob/main/CONTRIBUTING.md).

Hey,
I pushed another commit to fix the pre-commit errors. It seems that the tests were passed except for codecov/patch. I am not really sure what is the issue, can you guide me?

Thanks

**Comment by Cube707** [comment]
The _coverage/patch_ diplays the coverage of your added lines, which is currently 83.3%. This means that not all of the lines added by you are coverd by allready existing tests. This is to be expected, as you added branches to the code (new `if` statements, etc.) but didn't add tests for these cases.

To increase the overage, simply add the missing tests.

**Comment by Cube707** [comment]
for a detailed breakdown, see [the codecov webpage](https://app.codecov.io/gh/magmax/python-inquirer/pull/327). As you can see there, your added ìf`'s are marked yellow, as only on one of the possible code branches is codered (the one thats skips the if block is missing).

There is also a existing problem further up, where a line is not covered at all (maked red), you can ignore it but we missed it on a code review which is exacly what this test are here to prevent as much as possible

File: src/inquirer/render/console/_checkbox.py [hunk 1]
```diff
@@ -10,9 +10,15 @@
 class Checkbox(BaseConsoleRender):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.selection = [k for (k, v) in enumerate(self.question.choices) if v in (self.question.default or [])]
+        self.locked = self.question.locked or []
+        self.selection = [k for (k, v) in enumerate(self.question.choices) if v in (self.set_default_choices())]
```
Comment (Cube707) [inline src/inquirer/render/console/_checkbox.py:14]: remove redundent parantases

File: src/inquirer/render/console/_checkbox.py [hunk 2]
```diff
@@ -10,9 +10,15 @@
...
         self.current = 0
 
+    def set_default_choices(self):
+        default = self.question.default or []
+        final = set(default + self.locked)
+        return list(final)
```
Comment (Cube707) [inline src/inquirer/render/console/_checkbox.py:20]: why do you cast into a set and than back into a list? This semse redundant and bad-practise.

`in` works just as well on a set or a list with doubles in it..

**Review by Cube707 — CHANGES_REQUESTED** [review]
I think it should be visually more obvious that choises are locked. For example by changing the color (maybe a gray?)

File: src/inquirer/render/console/_checkbox.py (hunk 2 above)
Comment (NivEz) [reply src/inquirer/render/console/_checkbox.py]: fixed

**Comment by NivEz** [comment]
> I think it should be visually more obvious that choises are locked. For example by changing the color (maybe a gray?)

Good point, added gray color

File: src/inquirer/themes.py [hunk 3]
```diff
@@ -93,6 +93,7 @@ def __init__(self):
         self.Checkbox.selected_color = term.yellow + term.bold
         self.Checkbox.unselected_color = term.normal
         self.Checkbox.unselected_icon = "[ ]"
+        self.Checkbox.locked_option_color = term.gray50
```
Comment (Cube707) [inline src/inquirer/themes.py:96]: you forgot to also add it to the base class (Line 78)

File: tests/acceptance/test_checkbox.py [hunk 4]
```diff
@@ -147,3 +147,27 @@ def setUp(self):
...
+        self.sut = pexpect.spawn("python examples/checkbox_locked.py")
+        self.sut.expect("DevOps.*", timeout=1)
+
+    def test_default_selection(self):
+        self.sut.send(key.ENTER)
+        self.sut.expect(r"{'courses': \['Programming fundamentals'\]}", timeout=1)
+
+    def test_locked_option(self):
+        self.sut.send(key.SPACE)
+        self.sut.send(key.LEFT)
```
Comment (Cube707) [inline tests/acceptance/test_checkbox.py:164]: Its a good idea to test for both option to select a potion.

But in this case the test would actually work even if the locking code didn work, as it deselects and than imediatly selects againt. So please seperate this into two seperate tests

File: tests/integration/console_render/test_checkbox.py [hunk 5]
```diff
@@ -324,3 +324,59 @@ def test_double_invert_all(self):
         result = sut.render(question)
 
         assert result == []
+
+    def test_unselect_locked(self):
+        stdin_array = [key.SPACE, key.LEFT, key.ENTER]
```
Comment (Cube707) [inline tests/integration/console_render/test_checkbox.py:329]: same as above

File: src/inquirer/render/console/_checkbox.py [hunk 6]
```diff
@@ -10,9 +10,14 @@
 class Checkbox(BaseConsoleRender):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
-        self.selection = [k for (k, v) in enumerate(self.question.choices) if v in (self.question.default or [])]
+        self.locked = self.question.locked or []
+        self.selection = [k for (k, v) in enumerate(self.question.choices) if v in self.set_default_choices()]
         self.current = 0
 
+    def set_default_choices(self):
```
Comment (Cube707) [inline src/inquirer/render/console/_checkbox.py:17]: this function is named strangely. I assume you wanted two make its return type clear?

However like this its named as a _Setter_-Methode, which it is clearly not. 
```suggestion
    def default_choices(self):
```

**Review by Cube707 — CHANGES_REQUESTED** [review]
Just a thought: should it really be assumed that locked choise must be selected? wouldn't it also make senese to lock choises as they are and use the `default` setting to choose if they are selected or not. As its now the functionalaty overlapps and its not possible to display options that are no longer choosable based on previous inputs (I mean you could just remove these chises all together, but if we want to lock stuff this is a valid usecase??)

---

also it would be great to make the commits more atomic and have less small commits that just fix stuff from a previous commit. You can do this by resetting to a previous commit and _appending_  to it, than cherry pick the other commits. You can force-push it to your branch and the PR will update.

This is not mandatory and just a preference of mine to have clean commits that contain all changes belonging to a feature. For example: one commit for the feature, one commit for the tests and one commit for the documentations

**Comment by Cube707** [comment]
I also just fixed the security issue on main, so please rebase

File: tests/acceptance/test_checkbox.py (hunk 4 above)
Comment (NivEz) [reply tests/acceptance/test_checkbox.py:168]: Seprated into two tests

File: src/inquirer/render/console/_checkbox.py (hunk 6 above)
Comment (NivEz) [reply src/inquirer/render/console/_checkbox.py]: Renamed

File: tests/integration/console_render/test_checkbox.py (hunk 5 above)
Comment (NivEz) [reply tests/integration/console_render/test_checkbox.py]: Seprated into two tests as well

**Comment by NivEz** [comment]
@Cube707 
Didn't have much time since I was very busy. I fixed the comments.

Thanks for reviewing :)

**Review by Cube707 — APPROVED** [review]

**Comment by Cube707** [comment]
@NivEz sorry it took so long, but I finally had the time to look at this and merge it. Hope it can still be helpfull to you

**Pull request merged**